### PR TITLE
Fixes #37 - Buffers in multiple groups is now supported

### LINKED ISSUE #37: Always selects first group from centaur-tabs-buffer-groups-function
`centaur-tabs-buffer-groups-function` should return a list of groups but it doesn't help when the list has more than one group. It just considers the first group from the list. My `centaur-tabs-buffer-groups-function` is:

```
(defun compro/centaur-tabs-buffer-groups ()
    (let* ((project-root (projectile-project-p))
           (project-name (projectile-project-name))
           (buf (buffer-name))
           (groups
            (-filter
             (lambda (it) (not (null it)))
             (list
              (when (or (string-prefix-p "*sent mail" buf)
                        (string-prefix-p "*notmuch" buf)
                        (string-equal "*unsent mail*" buf))
                "Mail")
              (when (or (buffer-file-name)
                        (eq major-mode 'dired-mode))
                "Files")
              (when (eq major-mode 'helpful-mode)
                "Helpful")
              (when (or (eq major-mode 'telega-chat-mode)
                        (eq major-mode 'telega-root-mode))
                "Telegram")
              (when (string-prefix-p "*" buf)
                "Internal")
              (when (not (null project-root))
                (concat "Project: " project-name))))))
      (when (null groups)
        (push "Unregistered" groups))
      groups))
```

This will put a file based buffer in two groups: `File` and `Project: <project-name>` group. But the function which uses the above function is only based on the first element of the list. So, `centaur-tabs-counsel-switch-group` is of little use.

Basically, I sometimes want to see a list of open files while sometimes I want to see a list of open files in the current project. There is some bug which doesn't let this happen.

### PR DESCRIPTION
**Pull request by Compro-Prasad**
Switching groups will now show buffers from that group only, even if a buffer is
present in multiple groups. Previously this would fail to work if a buffer is in
multiple groups.

### CODE REVIEW

File: centaur-tabs.el [hunk 1]
```diff
@@ -1791,7 +1800,7 @@ Run as `centaur-tabs-quit-hook'."
   (interactive)
   (centaur-tabs-select-beg-tab t))
 
-(defun centaur-tabs-select-beg-tab (&optional backward _type)
```
Comment (Compro-Prasad) [inline centaur-tabs.el]: I removed `_type` because it wasn't being used in the function.

**Comment by Compro-Prasad** [comment]
Hello @ema2159 could you please test this?
I am currently using https://github.com/Compro-Prasad/emacs-config/blob/4579394a57f14da1efdb7250222f7ee8e15e1477/init.el#L862-L886 as a group function.

**Comment by ema2159** [comment]
Yes, give me just a little time, I've been busy so I've just been able to check minor issues!

**Comment by ema2159** [comment]
Sorry, the thing is I caught a throat infection so I've been in bed and I've only been able to answer stuff through my phone

**Comment by Compro-Prasad** [comment]
No problem. I think I am too excited about taking `centaur-tabs` to the next level.

**Comment by ema2159** [comment]
> No problem. I think I am too excited about taking `centaur-tabs` to the next level.

I sincerely appreciate your help. You've helped me a lot with your issues/PRs.

**Comment by Compro-Prasad** [comment]
Please take a look here. This adds a feature that was supposed to be there but it isn't. If we can have a list of groups representing a buffer, we should also be able to select any group and the group tabs display logic should be correct. Without this PR, many users wouldn't get this feature which is a very good addition. Consider it like, you can use mode groups and project groups at the same time. No need to change the group function everytime.

**Comment by ema2159** [comment]
Sorry, I forgot to try out this PR. I like the idea, but I encountered one issue, I move a lo with tab forward/backward movements (I have the cycle-scope variable set to nil). The thing is that, with this sometimes I'm unable to jump to another group with tab forward/backward movements, do you know why this is?

**Comment by Compro-Prasad** [comment]
I actually don't use that feature. When cycling we should change the window parameter which sets the group for current window. This should probably fix the issue.

**Comment by Compro-Prasad** [comment]
Still busy. Will find a time slot to work on this.

**Comment by Compro-Prasad** [comment]
There are some bugs that I found out while reading the code. I want to mention them here and fix when I get time.
1. `centaur-tabs-selected-tab` function returns the first selected tab, not the current tab in current window. This might be a bug. Try opening same group in different windows having different selected tabs. Should probably be fixed using `current-buffer` or `window-parameter` (if `current-buffer` is unable to solve it).
2. `centaur-tabs-cycle` tries to do too much. I want to break it into three functions for the three conditions. Wrapping will be enabled by default. Configuration will be provided to disable wrapping.
  1. `centaur-tabs-cycle-tabs` to cycle only tabs.
  2. `centaur-tabs-cycle-groups` to cycle only groups. If one group, cycle tabs.
  3. `centaur-tabs-cycle-default` to cycle tabs first then groups if we hit the edge.
3. Remove `centaur-tabs-tab-next` function as it is too simple(doesn't have any knowledge about current context).

**Comment by Compro-Prasad** [comment]
I have looked into this. And it seems that it is going to be a lot of work. I will have to update how groups are mapped to buffers.
I don't understand the need to update groups when switching tabs ( `centaur-tabs-current-tabset` accepts a `update` arg which seems to be useless as whenever this function is called, it is always passed a `t` arg which updates a lot of stuff by calling `centaur-tabs-buffer-update-groups` function. When switching tabs, this function gets called twice with `t` parameter passed even though it isn't required.). The sole problem is caused by the implementation of `centaur-tabs-buffer-update-groups`. I will have to isolate the situations, when to update and when not to update.

**Comment by Compro-Prasad** [comment]
Pull and check if it works.

File: centaur-tabs.el [hunk 2]
```diff
@@ -866,53 +914,18 @@ The scope of the cyclic navigation through tabs is specified by the
...
-		  tab (car (if backward (last tabset) tabset))))
-	  ;; Select the first/last visible tab of the new group.
-	  (setq tabset (centaur-tabs-tabs (centaur-tabs-tab-tabset tab))
-		tab (car (if backward (last tabset) tabset))))
-	))
-      (centaur-tabs-buffer-select-tab tab))))
+  (let (;; If navigation through groups is requested, and there is
+	;; only one group, navigate through visible tabs.
+	(cycle
+	 (if (and (eq centaur-tabs-cycle-scope 'groups)
```
Comment (Compro-Prasad) [inline centaur-tabs.el]: Why is this needed? IMO if an user prefers switching between groups then why should change the variable internally?

**Comment by ema2159** [comment]
There's a lot of things from the code that are from Aquamacs's tabs implementation. We have to be really really careful making changes this big. They may cause unexpected bugs.

**Comment by ema2159** [comment]
Also, I need you to use Flycheck to verify there's no linting and compiling errors when submitting your PR's! You usually submit pretty nice and ordered code so let's try to keep it that way. I have incredibly high coding standards so let's try to eliminate all possible warnings and keep the nice little green check in the mode-line (if not, my OCD will be heavily triggered jaja).

**Comment by Compro-Prasad** [comment]
That is why I asked you to test it properly before merging. After all this is my first package contribution that is this big. Besides, I have fixed what you asked for. Also, this fixes some edge cases that weren't handled before. There is a bug that I am unable to fix. This PR isn't a reason for the bug though. It wasn't there in the first place. When I fixed this multiple grouping issue, this new bug appeared. It wont affect users who don't use shared groups per buffer.
The bug: When you open two same buffers in different windows having different groups, the modified marker doesn't sync between them.

**Comment by ema2159** [comment]
I found a little bug. When doing `centaur-tabs-group-by-projectile-project` or `centaur-tabs-group-buffer groups`, the tabs should be updated and displayed immediately but for some reason they get the update after another centaur tabs command is inputed

**Comment by ema2159** [comment]
@Compro-Prasad don't worry,  you're doing a pretty good job, you've helped me a lot testing and fixing a lot of bugs with the package and I really appreciate your help. I'll try to test this thoroughly to ensure that the package keeps it's quality and I'll inform you about all the bugs I encounter!

**Comment by Compro-Prasad** [comment]
The one bug remaining is currently not possible to fix and isn't a big issue. Its just modified marker not getting synced. I will try fixing that another time. For now, we need to find out other bugs.

**Comment by ema2159** [comment]
@Compro-Prasad cool. It seems that it's working fine now and all the linting issues are gone (almost, two minor issues remain). I'll try to test it thoroughly and when I feel it's alright I'll merge it. Thanks for all your great help and enthusiasm with the package!.

**Comment by Compro-Prasad** [comment]
> By the way, just out of curiosity, where are you from?

India.

**Comment by ema2159** [comment]
> > By the way, just out of curiosity, where are you from?
> 
> India.

Awesome! It's been pretty nice to work with you

File: centaur-tabs.el (hunk 2 above)
Comment (Compro-Prasad) [reply centaur-tabs.el]: @ema2159 I asked this question to you. Is there anything you have to say about this?

**Comment by ema2159** [comment]
Hi @Compro-Prasad! Are you still using this functionality? I'm currently thinking on reverting this given that as far as I know, it's not very used by other users and has caused some unexpected bugs like #62 and #58. Please let me know what you think.

**Comment by Compro-Prasad** [comment]
I am using `tab-line` from Emacs 27. So, its okay to remove this feature.
